src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3284,71 +3309,392 @@ new_readdir(DIR *dirp)
+			goto err_out1;
+		}
+		new_envp[i] = new_preload_str;
+		i++;
+	}
+	if (!report_included) {
+		if (report)
+			str_report = strndup("D_IL_REPORT=1", 14);
+		else
+			str_report = strndup("D_IL_REPORT=0", 14);
+		if (str_report == NULL) {
+			printf("Error: failed to allocate memory for D_IL_REPORT env!\n");
+			goto err_out2;
+		}
+		new_envp[i] = str_report;
+		i++;
+	}
+	if (!mp_included) {
+		fs_root = getenv("DAOS_MOUNT_POINT");
+		if (fs_root) {
+			rc = asprintf(&str_mp, "DAOS_MOUNT_POINT=%s", fs_root);
+			if (rc < 0) {
+				printf("Error: failed to allocate memory for DAOS_MOUNT_POINT "
+				       "env!\n");
+				goto err_out3;
+			}
+			new_envp[i] = str_mp;
+			i++;
+		}
+	}
+	if (!pool_included) {
+		pool = getenv("DAOS_POOL");
+		if (pool) {
+			rc = asprintf(&str_pool, "DAOS_POOL=%s", pool);
+			if (rc < 0) {
+				printf("Error: failed to allocate memory for DAOS_MOUNT_POINT "
+				       "env!\n");
+				goto err_out4;
+			}
+			new_envp[i] = str_pool;
+			i++;
+		}
+	}
+	if (!cont_included) {
+		container = getenv("DAOS_CONTAINER");
+		if (container) {
+			rc = asprintf(&str_cont, "DAOS_CONTAINER=%s", container);
+			if (rc < 0) {
+				printf("Error: failed to allocate memory for DAOS_CONTAINER "
+				       "env!\n");
+				goto err_out5;
+			}
+			new_envp[i] = str_cont;
+			i++;
+		}
+	}
+	if (!maxeq_included) {
+		rc = asprintf(&str_maxeq, "D_IL_MAX_EQ=%d", eq_count_max);
+		if (rc < 0) {
+			printf("Error: failed to allocate memory for D_IL_MAX_EQ env!\n");
+			goto err_out6;
+		}
+		new_envp[i] = str_maxeq;
+		i++;
+	}
+	if (!enforcement_included) {
+		if (enforce_exec_env)
+			str_enforcement = strndup("D_IL_ENFORCE_EXEC_ENV=1", 24);
+		else
+			str_enforcement = strndup("D_IL_ENFORCE_EXEC_ENV=0", 24);
+		if (str_enforcement == NULL) {
+			printf("Error: failed to allocate memory for D_IL_ENFORCE_EXEC_ENV env!\n");
+			goto err_out7;
+		}
+		new_envp[i] = str_enforcement;
+		i++;
+	}
+
+	/* NULL pointer to end */
+	new_envp[i] = NULL;
+
+	return new_envp;
+
+err_out7:
+	free(str_maxeq);
+err_out6:
+	free(str_cont);
+err_out5:
+	free(str_pool);
+err_out4:
+	free(str_mp);
+err_out3:
+	free(str_report);
+err_out2:
+	free(new_preload_str);
+err_out1:
+	free(new_envp);
+err_out0:
+	return (char **)envp;
+}
+
+/* check whether fd 0, 1, and 2 are located on DFS mount or not. If yes, reopen files and
+ * set offset.
+ */
+static void
+setup_fd_0_1_2(void)
+{
+	int   i, fd, idx, fd_tmp, fd_new, open_flag;
+	off_t offset;
+
+	if (num_fd_dup2ed == 0)
+		return;
+
+	for (i = 0; i < MAX_FD_DUP2ED; i++) {
+		/* only check fd 0, 1, and 2 */
+		if (fd_dup2_list[i].fd_src >= 0 && fd_dup2_list[i].fd_src <= 2) {
+			fd        = fd_dup2_list[i].fd_src;
+			idx       = fd_dup2_list[i].fd_dest - FD_FILE_BASE;
+			offset    = file_list[idx]->offset;
+			open_flag = file_list[idx]->open_flag;
+
+			/* get a real fd from kernel */
+			fd_tmp = libc_open(file_list[idx]->path, open_flag);
```

Comment:
Do you have any consideration for if the path has changed, this could legitimately return a different file or ENOENT I assume?

Reply:
The code is moved to dup2(). If libc_open() fails, then dup2() fails with libc_open()'s errno.